src/utils.rs

```diff
@@ -0,0 +1,98 @@
+use crate::{CreatePsbtError, Input};
+use alloc::vec::Vec;
+use miniscript::bitcoin::{
+    absolute::{self, LockTime},
+    transaction::Version,
+    Sequence, Transaction, WitnessVersion,
+};
+
+use rand_core::{OsRng, RngCore};
+
+/// Applies BIP326 anti‐fee‐sniping
+pub fn apply_anti_fee_sniping(
+    tx: &mut Transaction,
+    inputs: &[Input],
+    current_height: absolute::Height,
+    rbf_enabled: bool,
+) -> Result<(), CreatePsbtError> {
+    const MAX_SEQUENCE_VALUE: u32 = 65_535;
+    const USE_NLOCKTIME_PROBABILITY: u32 = 2;
+    const MIN_SEQUENCE_VALUE: u32 = 1;
+    const FURTHER_BACK_PROBABILITY: u32 = 10;
+    const MAX_RANDOM_OFFSET: u32 = 99;
+
+    let mut rng = OsRng;
+
+    if tx.version < Version::TWO {
+        return Err(CreatePsbtError::UnsupportedVersion(tx.version));
+    }
+
+    let taproot_inputs: Vec<usize> = (0..tx.input.len())
+        .filter(|&idx| {
+            // Check if this input is taproot using the corresponding Input data
+            inputs
+                .get(idx)
+                .and_then(|input| input.plan())
+                .and_then(|plan| plan.witness_version())
+                .map(|version| version == WitnessVersion::V1)
+                .unwrap_or(false)
+        })
+        .collect();
+
+    // Check always‐locktime conditions
+    let must_use_locktime = inputs.iter().any(|input| {
+        let confirmation = input.confirmations(current_height);
+        confirmation == 0
+            || confirmation > MAX_SEQUENCE_VALUE
+            || !matches!(
+                input.plan().and_then(|plan| plan.witness_version()),
+                Some(WitnessVersion::V1)
+            )
```

Comment:
I was thinking it might be simpler and more robust to test if the input is taproot by using `Script::is_p2tr`, rationale being that we always have a prev txout but may not have a plan.

```suggestion
            || !input.prev_txout().script_pubkey.is_p2tr()
```

Reply:
good, this approach is much better.